spec/sql_database_spec.rb

```diff
@@ -0,0 +1,98 @@
+# frozen_string_literal: true
+
+require 'sql_database'
+
+RSpec.describe SQLDatabase do
+  let(:database) { described_class.new('file::memory:?cache=shared') }
+  let(:sql) { SQLite3::Database.open('file::memory:?cache=shared') }
+  let(:weather_data) { JSON.load_file('fixtures/london_weather.json', { symbolize_names: true }) }
+
+  before { database }
+  after { sql.execute 'DROP TABLE weather' }
+
+  describe '#augment_database' do
+    it 'adds weather information to the database' do
+      database.augment_database(weather_data)
+
+      result = sql.query "SELECT * FROM weather WHERE name = 'London'" do |rows|
+        rows.next_hash.transform_keys(&:to_sym)
+      end
+
+      expect(result).to eq(london_weather_output)
+    end
+
+    it 'updates the database if the same weather information already exists' do
+      database.augment_database(weather_data)
+      database.augment_database(weather_data)
+
+      result = sql.query "SELECT * FROM weather WHERE name = 'London'"
+
+      expect(result.count).to eq(1)
+    end
+
+    it 'returns a success message if no expections are made' do
+      result = database.augment_database(weather_data)
+
+      expect(result).to include('Successful')
+    end
+
+    context 'if data is bad' do
+      bad_data = { weather: [{ description: 'Irrelivent' }], main: {} }
+
+      it 'does not raise an error' do
+        expect { database.augment_database(bad_data) }.not_to raise_error
+      end
+      it 'returns a failure message' do
+        result = database.augment_database(bad_data)
+
+        expect(result).to include('Bad weather data! NOT NULL constraint failed')
+      end
+    end
+  end
+
+  describe '#retrieve_weather' do
+    it 'returns weather information matching city name passed' do
+      add_weather_row
+
+      result = database.retrieve_weather('London')
+
+      expect(result).to eq(london_weather_output)
+    end
+
+    it 'returns an empty hash if no matching city name is found' do
+      add_weather_row
+
+      result = database.retrieve_weather('Irrelivent')
+
+      expect(result).to eq({})
+    end
+  end
+
+  def london_weather_output # rubocop:disable Metrics/MethodLength
+    {
+      id: 1,
+      name: 'London',
+      unix_date: 1_617_973_201,
+      description: 'overcast clouds',
+      temp: 13.04,
+      feels_like: 11.63,
+      temp_min: 11.67,
+      temp_max: 14.44,
+      humidity: 47
+    }
+  end
+
+  def add_weather_row
+    sql.execute('INSERT INTO weather(
+    name, unix_date, description, temp, feels_like, temp_min, temp_max, humidity
+    ) VALUES(?, ?, ?, ?, ?, ?, ?, ?)', order_row(weather_data))
+  end
+
+  def order_row(data)
+    row = data.values_at(:name, :dt)
+    row << data[:weather].first[:description]
+    row << data[:main].values_at(
+      :temp, :feels_like, :temp_min, :temp_max, :humidity
+    )
+  end
```

Comment:
Since this is the test, instead of copying the same code I would have the explicit data:

```ruby
def order_row
  ['London', 1617973201, 'overcast clouds', [13.04, 11.63, 11.67, 14.44, 47]]
end
```

Reply:
Whoops! I ment to `+` in row 94